smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/serde/XmlParserGenerator.kt

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package software.amazon.smithy.kotlin.codegen.rendering.serde
+
+import software.amazon.smithy.codegen.core.Symbol
+import software.amazon.smithy.kotlin.codegen.core.KotlinWriter
+import software.amazon.smithy.kotlin.codegen.core.RuntimeTypes
+import software.amazon.smithy.kotlin.codegen.core.withBlock
+import software.amazon.smithy.kotlin.codegen.model.knowledge.SerdeIndex
+import software.amazon.smithy.kotlin.codegen.rendering.protocol.ProtocolGenerator
+import software.amazon.smithy.kotlin.codegen.rendering.protocol.toRenderingContext
+import software.amazon.smithy.model.shapes.MemberShape
+import software.amazon.smithy.model.shapes.OperationShape
+import software.amazon.smithy.model.shapes.Shape
+import software.amazon.smithy.model.shapes.StructureShape
+import software.amazon.smithy.model.traits.TimestampFormatTrait
+
+/**
+ * XML parser generator based on common deserializer interface and XML serde descriptors
+ */
+open class XmlParserGenerator(
+    // FIXME - shouldn't be necessary but XML serde descriptor generator needs it for rendering context
+    private val protocolGenerator: ProtocolGenerator,
+    private val defaultTimestampFormat: TimestampFormatTrait.Format
+) : StructuredDataParserGenerator {
+
+    open fun descriptorGenerator(
+        ctx: ProtocolGenerator.GenerationContext,
+        shape: Shape,
+        members: List<MemberShape>,
+        writer: KotlinWriter
+    ): XmlSerdeDescriptorGenerator = XmlSerdeDescriptorGenerator(ctx.toRenderingContext(protocolGenerator, shape, writer), members)
+
+    override fun operationDeserializer(
+        ctx: ProtocolGenerator.GenerationContext,
+        op: OperationShape,
+        members: List<MemberShape>
+    ): Symbol {
+        val outputSymbol = ctx.symbolProvider.toSymbol(ctx.model.expectShape(op.output.get()))
+        return op.bodyDeserializer(ctx.settings) { writer ->
+            addNestedDocumentDeserializers(ctx, op, writer)
+            val fnName = op.bodyDeserializerName()
+            writer.openBlock("private fun #L(builder: #T.Builder, payload: ByteArray) {", fnName, outputSymbol)
+                .call {
+                    renderDeserializeOperationBody(ctx, op, members, writer)
+                }
+                .closeBlock("}")
+        }
+    }
```

Comment:
**Question**: There's a fair bit of duplication between the JSON and XML implementations here. Is it worth considering format-agnostic generator superclasses for parsers/serializers?

Reply:
Only because they share a similar codepath today and use the common `Serializer` and `Deserializer` interfaces. If they were implemented e.g. directly off tokenization this would cease to be true. 


I am hoping we still have time to look into this at some point which I think would make the generated code easier to debug in real world scenarios. Today it works fine but it's hard to set breakpoints and understand what is happening. Perhaps when we get to KMP XML it's something we could consider...